File: app/warmup.py
"""Shared startup warmup status for health checks and chat readiness gates."""
from __future__ import annotations

import asyncio
import logging
from typing import Any

from app.config import settings


WARMUP_STATUS: dict[str, Any] = {
    "tts": {"status": "not_started", "backend": None, "error": None},
    "flashhead": {
        "status": "not_started",
        "worker_ready": False,
        "avatar_image_path": None,
        "error": None,
        "inference_warmup": {
            "status": "not_started",
            "elapsed_sec": 0.0,
            "error": None,
        },
    },
}


def reset_flashhead_warmup(image_path: str) -> None:
    WARMUP_STATUS["flashhead"].update(
        {
            "status": "started",
            "worker_ready": True,
            "avatar_image_path": image_path,
            "error": None,
            "inference_warmup": {
                "status": "not_started",
                "elapsed_sec": 0.0,
                "error": None,
            },
        }
    )


async def run_flashhead_warmup(image_path: str) -> None:
    """Run avatar and inference warmup, updating shared status in-place."""
    from app.services.flashhead.persistent import worker_manager

    logger = logging.getLogger(__name__)
    reset_flashhead_warmup(image_path)
    try:
        worker_manager.start()
        WARMUP_STATUS["flashhead"]["worker_ready"] = True
    except Exception as exc:  # noqa: BLE001 - warmup status carries the failure.
        WARMUP_STATUS["flashhead"].update(
            {
                "status": "failed",
                "worker_ready": False,
                "error": str(exc),
            }
        )
        WARMUP_STATUS["flashhead"]["inference_warmup"].update(
            {"status": "failed", "elapsed_sec": 0.0, "error": str(exc)}
        )
        logger.warning("FlashHead worker start failed avatar=%s error=%s", image_path, exc)
        return
    try:
        await worker_manager.warmup_background(image_path)
        WARMUP_STATUS["flashhead"]["status"] = "ok"
    except Exception as exc:  # noqa: BLE001 - warmup status carries the failure.
        WARMUP_STATUS["flashhead"]["status"] = "failed"
        WARMUP_STATUS["flashhead"]["error"] = str(exc)
        WARMUP_STATUS["flashhead"]["inference_warmup"].update(
            {"status": "failed", "elapsed_sec": 0.0, "error": str(exc)}
        )
        logger.warning("FlashHead avatar warmup failed avatar=%s error=%s", image_path, exc)
        return

    inference = WARMUP_STATUS["flashhead"]["inference_warmup"]
    inference.update({"status": "started", "elapsed_sec": 0.0, "error": None})
    try:
        elapsed_sec = 0.0
        for duration_sec in (2.0, 3.2, 5.0):
            result = await worker_manager.warmup_inference_background(
                image_path,
                duration_sec=duration_sec,
                fps=float(settings.default_fps),
            )
            elapsed_sec += float(result.get("elapsed_sec") or 0.0)
        inference.update(
            {
                "status": "ok",
                "elapsed_sec": round(elapsed_sec, 3),
                "error": None,
            }
        )
    except Exception as exc:  # noqa: BLE001 - warmup status carries the failure.
        inference.update(
            {
                "status": "failed",
                "elapsed_sec": 0.0,
                "error": str(exc),
            }
        )
        logger.warning("FlashHead inference warmup failed avatar=%s error=%s", image_path, exc)


def start_flashhead_warmup_for_avatar(image_path: str) -> None:
    """Schedule FlashHead warmup from an async route or lifespan task."""
    reset_flashhead_warmup(image_path)
    asyncio.create_task(run_flashhead_warmup(image_path))


def is_digital_human_ready() -> bool:
    if settings.musetalk_backend != "local":
        return WARMUP_STATUS.get("tts", {}).get("status") == "ok"

    flashhead = WARMUP_STATUS.get("flashhead") or {}
    inference = flashhead.get("inference_warmup") or {}
    tts = WARMUP_STATUS.get("tts") or {}
    return (
        tts.get("status") == "ok"
        and flashhead.get("worker_ready") is True
        and flashhead.get("status") == "ok"
        and inference.get("status") == "ok"
    )

File: frontend/src/warmup.ts
import type { HealthPayload } from "./types";

export function isDigitalHumanReady(payload?: HealthPayload | null): boolean {
  const warmup = payload?.warmup || {};
  const tts = warmup.tts || {};
  const profile = payload?.profile || {};
  const mode = profile.deployment_mode || "mock";
  if (mode === "mock") return tts.status === "ok";
  if (mode === "remote" && profile.render_api_url) return tts.status === "ok";

  const flashhead = warmup.flashhead || {};
  const inference = flashhead.inference_warmup || {};
  return (
    tts.status === "ok" &&
    flashhead.worker_ready === true &&
    flashhead.status === "ok" &&
    inference.status === "ok"
  );
}

export function warmupFailureMessage(payload?: HealthPayload | null): string {
  const warmup = payload?.warmup || {};
  const tts = warmup.tts || {};
  const flashhead = warmup.flashhead || {};
  const inference = flashhead.inference_warmup || {};
  if (tts.status === "failed") return "语音服务初始化失败，请刷新重试";
  if (flashhead.status === "failed") return "数字人头像初始化失败，请刷新重试";
  if (inference.status === "failed") return "数字人推理初始化失败，请刷新重试";
  return "";
}

export function warmupProgressMessage(payload?: HealthPayload | null): string {
  if (!payload) return "正在连接后端服务...";
  const warmup = payload.warmup || {};
  const tts = warmup.tts || {};
  const profile = payload.profile || {};
  const mode = profile.deployment_mode || "mock";
  const flashhead = warmup.flashhead || {};
  const inference = flashhead.inference_warmup || {};

  if (tts.status !== "ok") {
    return tts.status === "started" ? "正在预热语音服务..." : "正在检查语音服务...";
  }
  if (mode === "mock") return "可以开始聊天";
  if (mode === "remote" && profile.render_api_url) return "可以开始聊天";
  if (flashhead.worker_ready !== true) return "正在启动数字人渲染进程...";
  if (flashhead.status !== "ok") {
    return flashhead.status === "started" ? "正在准备数字人头像..." : "正在检查数字人头像...";
  }
  if (inference.status !== "ok") {
    return inference.status === "started"
      ? "正在预热数字人推理，首次启动约需 1 分钟..."
      : "正在等待数字人推理预热...";
  }
  return "可以开始聊天";
}
